File: src/main/java/com/chatdemo/repository/ConversationRepository.java
package com.chatdemo.repository;

import com.chatdemo.model.ConversationMessage;
import dev.langchain4j.data.message.ChatMessage;

import java.util.List;

/**
 * Repository interface for conversation persistence.
 * Separates concerns between full history (for UI) and active messages (for LLM).
 */
public interface ConversationRepository {
    
    /**
     * Get full conversation history including archived messages.
     * Used for displaying to the user.
     */
    List<ConversationMessage> getFullHistory(String conversationId);
    
    /**
     * Get only active (non-archived) messages for the LLM.
     */
    List<ChatMessage> getActiveMessages(String conversationId);
    
    /**
     * Get the current summary of archived messages.
     * @return summary text, or null if no summarization has occurred
     */
    String getSummary(String conversationId);
    
    /**
     * Add a new message to the conversation.
     */
    void addMessage(String conversationId, ChatMessage message);
    
    /**
     * Archive messages that have been summarized.
     * Marks the specified messages as archived and updates the summary.
     */
    void archiveMessages(String conversationId, List<String> messageIds, String newSummary);
    
    /**
     * Clear all messages and summary for a conversation.
     */
    void clear(String conversationId);
}

File: src/main/scala/com/chatdemo/backend/repository/ConversationRepository.scala
package com.chatdemo.backend.repository

import com.chatdemo.common.model.{ConversationMessage, MessageAttachment}
import dev.langchain4j.data.message.ChatMessage

/**
 * Repository interface for conversation persistence.
 * Separates concerns between full history (for UI) and active messages (for LLM).
 */
trait ConversationRepository {

  /** Create a new conversation with the given ID. Returns true if created, false if it already exists. */
  def createConversation(conversationId: String): Boolean

  /** Check whether a conversation exists. */
  def conversationExists(conversationId: String): Boolean

  /** List all conversation IDs. */
  def listConversationIds(): List[String]

  /** Get full conversation history including archived messages. Used for displaying to the user. */
  def getFullHistory(conversationId: String): List[ConversationMessage]

  /** Get only active (non-archived) messages for the LLM. */
  def getActiveMessages(conversationId: String): List[dev.langchain4j.data.message.ChatMessage]

  /** Get the current summary of archived messages. */
  def getSummary(conversationId: String): String

  /** Add a new message to the conversation. */
  def addMessage(conversationId: String, message: dev.langchain4j.data.message.ChatMessage): Unit

  /** Attach metadata to the latest AI message in the conversation. */
  def attachToLatestAiMessage(conversationId: String, attachments: List[MessageAttachment]): Unit = {}

  /** Attach metadata to the latest User message in the conversation. */
  def attachToLatestUserMessage(conversationId: String, attachments: List[MessageAttachment]): Unit = {}

  /** Archive messages that have been summarized. */
  def archiveMessages(conversationId: String, messageIds: List[String], newSummary: String): Unit

  /** Clear all messages and summary for a conversation. */
  def clear(conversationId: String): Unit
}
